feat: Cache compiled routes

come-nc opened the pull request:
PR for https://github.com/nextcloud/server/issues/54155

* Follow-up of https://github.com/nextcloud/server/pull/52667

## Summary

1. Use symfony tools to compile all routes
2. Cache that in a local cache
3. Load from cache instead of parsing routes on each request
4. [x] ~Do the same for the generator~ postponed to follow-up if needed
5. [x] Check whether cache reset is needed in some situations

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

come-nc committed:
feat(router): Cache routes in local cache if possible

This is not ideal because serializing the routecollection is not easy.
It seems Symfony has its own way of doing things by dumping routes to a
 PHP file, maybe that would be better, but it would mean pulling a new
 symfony dependency and maybe refactor our Router.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
feat(router): Use Symfony CompiledUrlMatcherDumper to cache routes

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix: Use a more fitting cache key

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
Notes:

1. Big win on core routes like login. ~100ms saved
2. Smaller win on apps/ routes, ~20ms saved
3. On apps/ routes, 2MB more memory usage, could be avoided by caching by app, but then we lose part of the time improvement
4. We lose routes pointing to a closure, because that cannot be serialized. This is not used a lot and mainly used for including files throuh IRoute::actionInclude. We should deprecate that and add a work-around, but I’m short on ideas, we might have to load the routes again for these ones (if we cache in which file they are it might not be that expensive).

come-nc committed:
fix: Deprecate action and actionInclude from routes and fix actionInclude

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
feat(router): Add fallback for legacy action and file routes

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix: Fix psalm issue by adding missing parameter for profiler

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
> 4. [ ]  Do the same for the generator

I am not sure if it makes sense for generations, as that’s used less often I suppose?
We could still implement in a follow-up if needed.

> 5. [ ]  Check whether cache reset is needed in some situations

Cache has a prefix using applications versions, so we’re fine on this point.

come-nc commented:
Ready for reviews. Should not be merged yet so that we can compare with master performances more easily.

provokateurin commented on lib/private/Route/CachingRouter.php:
```diff
@@ -54,4 +55,62 @@ public function generate($name, $parameters = [], $absolute = false) {
+				$this->loaded = true;
+				$this->eventLogger->end('cacheroute:load:' . $app);
+				return;
+			}
```
Shouldn't this get skipped (early-return) if `this->loaded === true` similar to how it is done with `$this->loadedApps`?

come-nc replied:
So, in the final version, I think I cannot do that.
Because to create a `CompiledUrlMatcher` I need the serialized version of the routes.
What I could do is `$cachedRoutes = $this->serializeRouteCollection($this->root);` if `$this->loaded` is true.
But is that faster/better than `$cachedRoutes = $this->cache->get($key);`? I do not think so.

Also findMatchingRoute is only called once in normal usage of the router.

provokateurin commented on lib/private/Route/CachingRouter.php:
```diff
@@ -54,4 +55,62 @@ public function generate($name, $parameters = [], $absolute = false) {
+			}
+		}
+		parent::loadRoutes($app);
+		$this->cache->set('root:' . $requestedApp, $this->serializeRouteCollection($this->root), 3600);
```
So the routes will automatically get invalidated if an app is enabled/disabled/updated because we use all app versions to build a global cache prefix in the background?

provokateurin commented on lib/private/Route/CachingRouter.php:
```diff
@@ -54,4 +60,98 @@ public function generate($name, $parameters = [], $absolute = false) {
+		$key = $this->context->getHost() . '#' . $this->context->getBaseUrl() . '#rootCollection';
+		$cachedRoutes = $this->cache->get($key);
+		if (!$cachedRoutes) {
+			parent::loadRoutes();
```
So for caching this method is called without any specific app. Wouldn't it make sense to remove the app parameter completely in order to simplify the logic behind it and also avoid having it potentially do unnecessary work by only loading one route at app at first and then everything again?

come-nc replied:
No, because if no local cache is available, Router is used, and we do not want to load all routes when in an app.
It’s only when we know that we can cache them that we load all routes no matter where we are.

provokateurin commented on lib/private/Route/Router.php:
How about making this Router abstract and removing the implementation of `findMatchingRoute`?
Or maybe even better, why not merge Router and CachingRouter?
I really dislike having the implementation split across two classes, while CachingRouter is clearly what should be always used.

come-nc replied:
Because local cache may not be available.

provokateurin reviewed (commented):
I reviewed commit-by-commit so it is a bit messy because some commits should have been squashed to make sense :see_no_evil:

ChristophWurst reviewed (approved):
Looks good! Didn't test

come-nc commented:
Failing CI is related.

Current theory:
- In some cases after enabling an application its routes are missing from cached routes
- This is supposed to be avoided by the cache prefix based on application versions, but that does not take into account disabled/enabled applications, only the version in DB

I will try to use only enabled applications for the cache prefix.

come-nc committed:
fix: Use only enabled applications versions in the cache prefix

This makes sure the cached routes are updated after enabling a
 previously disabled application

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
chore(tests): Adapt AppManager test to the use of searchValues

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix: Fix log display for integration tests

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
chore(tests): Add test for AppConfig::getAppInstalledVersions

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix: Use application ids as well as versions in the cache prefix

Avoids a corner case issue if one application was disabled and another
 one enabled with the same version, just to be safe.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix: Load all partially enabled apps routes

This avoids cache issues if some apps are only enabled for some groups,
 by loading the routes from all apps enabled for anyone, not only the
 current user.
If the queried application is disabled for the current user based on
 groups, the middleware will cancel the query.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
chore: update psalm baseline

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented on lib/private/Route/CachingRouter.php:
```diff
@@ -54,4 +60,98 @@ public function generate($name, $parameters = [], $absolute = false) {
+		if (!$cachedRoutes) {
+			parent::loadRoutes();
+			$cachedRoutes = $this->serializeRouteCollection($this->root);
+			$this->cache->set($key, $cachedRoutes, 3600);
```
No, the cache have a global prefix with all apps version numbers, see https://github.com/nextcloud/server/blob/acc2311a0d685281c54e8c6f1eecf7812253363a/lib/private/Server.php#L605
